### ISSUE TITLE
Multicursor (extremely) slow

### ISSUE THREAD

**Issue by alexandernst** [atom/atom#9871]
Using multicursor with more than 50-70 lines is extremely slow (running Atom in safe mode). When running with a theme (monokai) and a few basic plugins the thing becomes completely unusable when selecting more than 20-30 lines.

**Comment by nicekiwi** [comment]
+1 

Using it to turn a column from Excel into a coma separated list over 1000 lines does take a much longer amount of time than it should.

**Comment by shssoichiro** [comment]
Experienced this as well. Tried using multiple select to add the same text to the end of 3000 lines. Selecting the text and converting to multiple cursor mode worked okay. I typed one character and the editor froze.

This may or may not be related to the same issue, but using "Replace All" with a large number of replacements (in my tested case, 4000) also causes Atom to freeze.

**Comment by benogle** [comment]
cc @nathansobo

**Comment by nathansobo** [comment]
The fix is batching computation on edits and is in progress now. Should improve soon.

**Comment by alexandernst** [comment]
@nathansobo Sounds interesting. Is there some place I could seek for some more information about this improvement? (Or maybe you'll write a blog post when it's done?)

**Comment by alexandernst** [comment]
@nathansobo Any news on this one?

**Comment by nathansobo** [comment]
Working on it now. Reached a milestone last Friday in the ability to batch updates of intermediate spatial mapping structures but there's still work to do.

**Comment by MartinMuzatko** [comment]
Looking forward to this!

**Comment by alexandernst** [comment]
@nathansobo Any news on this?

**Comment by nathansobo** [comment]
@alexandernst Deprecating some APIs that are extremely harmful to multi-cursor performance is on our short term radar for the next 1-2 months. We have a batched text change API but we have yet to deprecate the non-batched version, which we plan to do soon. Last I checked on the profiles, the next big thing is streamlining our approach to auto-indent to be more batch-friendly. Our immediate focus has been getting large files on screen more quickly by gathering up the benefits of the display layer, and that's been going well. Multi-cursor is on the short list after that.

**Comment by adamreisnz** [comment]
@nathansobo any progress since your last update on improving the performance with multi-cursor?

I've ran into this as well (https://github.com/joseramonc/multi-cursor/issues/35) and find it very slow and unworkable when working with 100+ lines/cursors.

**Comment by nathansobo** [comment]
We unfortunately still haven't gotten to it.

**Comment by matonga** [comment]
Hi, it's been a long time to me suffering this issue. Since a few days I upgraded my software:

Ubuntu Gnome 17.10 -> Ubuntu 18.04 LTS
Atom v1.6.9 -> Atom v1.28.0 (+ atom-ide plugin)

And the issue is gone! Tested with 1000 and 2000 lines. I don't know which thing did the trick (Ubuntu version, Atom version, or the atom-ide plugin) so I'm posting this info here, just in case it may be useful to someone else.
